java-tools/OsmAndServer/src/main/java/net/osmand/server/api/services/OsmAndMapsService.java

```diff
@@ -771,28 +844,48 @@ private synchronized void initNewObfFiles(File target, File targetTemp) throws I
 		if (ref.fileIndex != null) {
 			ref.fileIndex = null;
 		}
-		if (ref.reader != null) {
-			ref.reader.close();
+		
+		if (!ref.readers.isEmpty()) {
+			for (Entry<BinaryMapIndexReader, Boolean> r : ref.readers.entrySet()) {
+				r.getKey().close();
+			}
 		}
 		if (nativelib != null) {
 			nativelib.closeMapFile(target.getAbsolutePath());
 		}
 		target.delete();
 		targetTemp.renameTo(target);
-		RandomAccessFile raf = new RandomAccessFile(target, "r"); //$NON-NLS-1$ //$NON-NLS-2$
-		ref.reader = new BinaryMapIndexReader(raf, target);
-		ref.fileIndex = cacheFiles.addToCache(ref.reader, target);
+		RandomAccessFile raf = new RandomAccessFile(target, "r");
+		BinaryMapIndexReader reader = new BinaryMapIndexReader(raf, target);
+		ref.readers.put(reader, true);
+		ref.fileIndex = cacheFiles.addToCache(reader, target);
 		cacheFiles.writeToFile(new File(config.cacheLocation, CachedOsmandIndexes.INDEXES_DEFAULT_FILENAME));
 		if (nativelib != null) {
 			nativelib.initMapFile(target.getAbsolutePath(), false);
 		}
 		LOGGER.info("Init new obf file " + target.getName() + " " + (System.currentTimeMillis() - val) + " ms");
 	}
 	
-	public synchronized BinaryMapIndexReader[] getObfReaders(QuadRect quadRect, List<LatLon> bbox) throws IOException {
+	public BinaryMapIndexReader[] getObfReaders(QuadRect quadRect, List<LatLon> bbox, int maxNumberMaps) throws IOException, InterruptedException {
```

Comment:
this file should return BinaryMapIndexReaderReference, to use pattern try(lock)/finally (unlock)